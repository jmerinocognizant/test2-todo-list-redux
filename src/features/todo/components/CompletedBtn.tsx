interface props {
    completed: boolean,
    onClickHandler: () => void,
}

export const CompletedBtn = ({completed,onClickHandler}:props) => {
  return (
    <>
        {
        completed?
            <button aria-label="btn-completed" className="btn btn-block btn-primary item-btn" onClick={onClickHandler}>
                <i className="fa-solid fa-check" ></i>
            </button>
            :
            <button aria-label="btn-completed" className="btn btn-block btn-outline-primary item-btn" onClick={onClickHandler}>
                <i className="fa-solid fa-check" ></i>
            </button>
        }    
    </>
  )
}
