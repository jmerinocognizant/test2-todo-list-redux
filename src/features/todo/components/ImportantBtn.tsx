interface props {
    important: boolean,
    onClickHandler: () => void,
}

export const ImportantBtn = ({important,onClickHandler}:props) => {
  return (
    <>
        {
        important?
            <button aria-label="btn-important" className="btn btn-block btn-primary item-btn" onClick={onClickHandler}>
                <i className="fa-solid fa-exclamation" ></i>
            </button>
            :
            <button aria-label="btn-important" className="btn btn-block btn-outline-primary item-btn" onClick={onClickHandler}>
                <i className="fa-solid fa-exclamation" ></i>
            </button>
        }    
    </>
  )
}
