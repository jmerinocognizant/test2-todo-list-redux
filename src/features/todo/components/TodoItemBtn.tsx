interface props {
    checked: boolean,
    classNameIcon: string,
    onClickHandler: () => void,
}

export const TodoItemBtn = ({checked,classNameIcon,onClickHandler}:props) => {
    const btnClassName = checked? "btn btn-block btn-primary item-btn":"btn btn-block btn-outline-primary item-btn"
    
  return (
    <>
         <button aria-label="btn-todo" className={btnClassName} onClick={onClickHandler}>
                <i className={classNameIcon} ></i>
         </button>
    </>
  )
}
