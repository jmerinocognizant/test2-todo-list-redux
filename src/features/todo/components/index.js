export * from './CompletedBtn';
export * from './ImportantBtn';
export * from './TodoAdd';
export * from './TodoItem';
export * from './TodoItemBtn';
export * from './TodoItemDescription';
export * from './TodoItemEditable';
export * from './TodoList';