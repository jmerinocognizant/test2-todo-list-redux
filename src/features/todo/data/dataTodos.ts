import { Todo, TodoState } from "../interfaces/interfaces";

export const todoList = [{
    id: 1,
    description: 'Install NodeJS',
    completed: true,
    important: true,
},{
    id: 2,
    description: 'Install Angular CLI',
    completed: true,
    important: false,
},{
    id: 3,
    description: 'Create new app',
    completed: false,
    important: true,
}, {
    id: 4,
    description: 'Serve app',
    completed: false,
    important: false,
}, {
    id: 5,
    description: 'Develop app',
    completed: false,
    important: false,
}, {
    id: 6,
    description: 'Deploy app',
    completed: false,
    important: false,
}]


export const initialState: TodoState = {
    todos: todoList,
    edit: 0,
  };

export const initialStateEmpty: TodoState = {
    todos: [],
    edit: 0,
  };

export const newTodo: Todo = {
    id: 10000,
    description: "new TODO Pruebas",
    completed: false,
    important: false,
}  