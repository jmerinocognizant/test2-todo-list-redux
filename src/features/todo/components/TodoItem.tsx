import { Todo } from "../interfaces/interfaces"
import { CompletedBtn } from "./CompletedBtn";
import { ImportantBtn } from "./ImportantBtn";
import { TodoItemDescription } from "./TodoItemDescription";
import { TodoItemEditable } from "./TodoItemEditable";
import { useAppDispatch } from '../../../app/hooks';
import {
    removeTodo,
    updateTodo,
    setEditable,
  } from '../store/todoSlice';
import { TodoItemBtn } from "./TodoItemBtn";
  
interface props {
    todo: Todo,
    editable: boolean,
}

export const TodoItem = ( {todo,editable}:props) => {
  
    const dispatch = useAppDispatch();

    const onDoubleClick = () => {
        dispatch(setEditable(todo.id));
    }

    const onRemoveTodo = () => {
        dispatch(removeTodo(todo.id));
    }

    const handleClickImportant = () => {
        const todoUpdated = {...todo,
                            important: !todo.important };
        dispatch(updateTodo(todoUpdated));
    }

    const handleClickCompleted = () => {
        const todoUpdated = {...todo,
            completed: !todo.completed };
        dispatch(updateTodo(todoUpdated));
    }

    const onSaveClickHandler = (textDescription:string) =>{
        const todoUpdated = {...todo,
            description: textDescription };
        dispatch(updateTodo(todoUpdated));
        dispatch(setEditable(0));
    }

    return (
        <>    
            < hr/>
            <div aria-label="todo-item" className="container-fluid todo-item">
                <div className="row">
                    <div className="col-auto mt-0  v-top">
                        {/* <CompletedBtn completed={todo.completed} onClickHandler={handleClickCompleted}/>  */}
                        <TodoItemBtn aria-label="btn-completed" checked={todo.completed} classNameIcon="fa-solid fa-check" onClickHandler={handleClickCompleted} />
                    </div>
                    <div className="col-auto mt-0  v-top">
                        {/* <ImportantBtn important={todo.important} onClickHandler={handleClickImportant}/> */}
                        <TodoItemBtn aria-label="btn-important" checked={todo.important} classNameIcon="fa-solid fa-exclamation" onClickHandler={handleClickImportant} />
                    </div>
                    
                    <div className="col">
                        {
                        editable ?(
                            <TodoItemEditable description = {todo.description} onSaveClick={onSaveClickHandler}/>  
                        ):(  
                            <TodoItemDescription description={todo.description} onDoubleClick={onDoubleClick}/>      
                        )
                        }
                    </div>
                    <div className="col-auto mt-0">
                        <button aria-label="btn-remove" className="btn btn-danger align-top" onClick={onRemoveTodo}>
                            remove
                        </button>
                    </div>
                </div>
            </div>
        </>
    )
}
