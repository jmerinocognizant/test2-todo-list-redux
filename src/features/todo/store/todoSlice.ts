import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState} from '../../../app/store';
import { todoList } from '../data/dataTodos';
import { Todo, TodoState } from '../interfaces/interfaces';

const initialState: TodoState = {
  todos: todoList,
  edit: 0,
};

export const todoSlice = createSlice({
  name: 'todo',
  initialState,
  // The `reducers` field lets us define reducers and generate associated actions
  reducers: {
    addTodo: (state, action: PayloadAction<Todo>) => {
      state.todos.push(action.payload);
    },
    removeTodo: (state, action: PayloadAction<number>) => {
      state.todos = state.todos.filter(item => item.id !== action.payload);
    },
    updateTodo: (state, action: PayloadAction<Todo>) => {
      const updatedTodo = action.payload;
      state.todos = state.todos.map(item => { 
        if(item.id === updatedTodo.id){
          return updatedTodo;
        }
        return item;
      });
    },
    setEditable: (state, action: PayloadAction<number>) => {
      state.edit = action.payload;
    },
  },

});

export const selectTodos = (state: RootState) => state.todo.todos;
export const selectEditable = (state: RootState) => state.todo.edit;

export const { addTodo, removeTodo, updateTodo, setEditable } = todoSlice.actions;

export default todoSlice.reducer;
