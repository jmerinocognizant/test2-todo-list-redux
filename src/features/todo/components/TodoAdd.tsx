import { useForm } from "../hooks/useForm";
import { Todo } from "../interfaces/interfaces";
import { useAppDispatch } from '../../../app/hooks';
import { addTodo } from '../store/todoSlice';

export const TodoAdd = () => {
  
  const dispatch = useAppDispatch();

  const { description, onInputChange, onResetForm } = useForm({
    description: ''
    });

  const onFormSubmit = ( event : React.FormEvent<HTMLFormElement> ) => {
    event.preventDefault();
    if ( description.length <= 1 ) return;

    const newTodo:Todo = {
        id: new Date().getTime(),
        description: description,
        completed: false,
        important: false,
    }

    dispatch(addTodo(newTodo));
    onResetForm();
  }
  
  return (
    <>
      <div className="container-fluid">
          <form aria-label="form" onSubmit = {onFormSubmit }>

          <div className="row">
            <div className="col">
              <input 
                      type="text" 
                      placeholder="Add something to do..."
                      className="form-control align-top"
                      name="description"
                      value={ description }
                      onChange={ onInputChange }
                  />
              </div>
            <div className="col-auto mt-0">      
              <button
                  type="submit" 
                  className="btn btn btn-primary align-top"
              >
                  Save
              </button>
            </div>
            </div>  
          </form>
      </div>
    </>
  )
}
