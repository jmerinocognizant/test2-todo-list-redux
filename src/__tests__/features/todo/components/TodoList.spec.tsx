import { fireEvent, render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { CompletedBtn, TodoAdd, TodoList } from '../../../../features/todo/components';
import configureMockStore from 'redux-mock-store';
import { selectTodos,selectEditable } from '../../../../features/todo/store/todoSlice';
import { useAppSelector } from '../../../../app/hooks';
import { store } from '../../../../app/store';
import { initialState, initialStateEmpty } from '../../../../features/todo/data/dataTodos';
import thunk from 'redux-thunk';

describe('Pruebas en <TodoList />', () => { 
    
    beforeEach( () => jest.clearAllMocks());

   //const mockStore = configureMockStore([]);
    
   // const {todos:returnTodos} = initialState;

    //const mockUseAppSelector =  jest.fn();
/*
    const mockUseAppSelector = jest.fn().mockReturnValueOnce(returnTodos).mockReturnValue(0);

    jest.mock('../../../../app/hooks', () =>({
        useAppSelector: mockUseAppSelector
    }));
*/
    
/*
    jest.mock('react-redux', () => ({
        useSelector: jest.fn().mockImplementation(selector => selector()),
      }));

    
    jest.mock('../../../../features/todo/store/todoSlice', () => ({
        selectTodos: jest.fn().mockReturnValue(todos),
        selectEditable: jest.fn().mockReturnValue(0),
    }));
*/

    test('debe de hacer match con el snapshot', () => {
        const middlewares = [thunk];
        const mockStore = configureMockStore(middlewares);
        const storeEmpty = mockStore({todo: initialState});
        const { container } = render(
            <Provider store={storeEmpty}>
                <TodoList />
            </Provider>
       );
       //screen.debug();
       expect( container ).toMatchSnapshot();
      
    });

    test('debe de mostrar un mensaje si la lista estÃ¡ vacia', () => {
        const middlewares = [thunk];
        const mockStore = configureMockStore(middlewares);
        const storeEmpty = mockStore({todo: initialStateEmpty});
        const noRecords = "No records";

        //console.log(storeEmpty.getState());
        
        const { container } = render(
            <Provider store={storeEmpty}>
                <TodoList />
            </Provider>
       );       
      
       //screen.debug();
       //console.log(storeEmpty);
       expect(screen.getByText(noRecords));
    });

    test('debe de mostrar los elementos de la lista', () => {
        const middlewares = [thunk];
        const mockStore = configureMockStore(middlewares);
        const storeLoaded = mockStore({todo: initialState});

        //console.log(storeEmpty.getState());
        const { container } = render(
            <Provider store={storeLoaded}>
                <TodoList />
            </Provider>
       );       
      
       //screen.debug();
       //console.log(storeEmpty);
       const todoItems = screen.getAllByLabelText('todo-item');
       //console.log(todoItems.length);
      expect(todoItems.length).toBe(initialState.todos.length);
    });

});