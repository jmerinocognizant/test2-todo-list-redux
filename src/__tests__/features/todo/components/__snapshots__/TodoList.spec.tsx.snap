// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoList /> debe de hacer match con el snapshot 1`] = `
<div>
  <div
    class="container-fluid"
  >
    <ul
      class="list-group mt-3"
    >
      <hr />
      <div
        aria-label="todo-item"
        class="container-fluid todo-item"
      >
        <div
          class="row"
        >
          <div
            class="col-auto mt-0  v-top"
          >
            <button
              aria-label="btn-todo"
              class="btn btn-block btn-primary item-btn"
            >
              <i
                class="fa-solid fa-check"
              />
            </button>
          </div>
          <div
            class="col-auto mt-0  v-top"
          >
            <button
              aria-label="btn-todo"
              class="btn btn-block btn-primary item-btn"
            >
              <i
                class="fa-solid fa-exclamation"
              />
            </button>
          </div>
          <div
            class="col"
          >
            <div
              aria-label="item-description"
              class="w-100 p-0"
            >
              <span
                class="h6 v-top "
              >
                Install NodeJS
              </span>
            </div>
          </div>
          <div
            class="col-auto mt-0"
          >
            <button
              aria-label="btn-remove"
              class="btn btn-danger align-top"
            >
              remove
            </button>
          </div>
        </div>
      </div>
      <hr />
      <div
        aria-label="todo-item"
        class="container-fluid todo-item"
      >
        <div
          class="row"
        >
          <div
            class="col-auto mt-0  v-top"
          >
            <button
              aria-label="btn-todo"
              class="btn btn-block btn-primary item-btn"
            >
              <i
                class="fa-solid fa-check"
              />
            </button>
          </div>
          <div
            class="col-auto mt-0  v-top"
          >
            <button
              aria-label="btn-todo"
              class="btn btn-block btn-outline-primary item-btn"
            >
              <i
                class="fa-solid fa-exclamation"
              />
            </button>
          </div>
          <div
            class="col"
          >
            <div
              aria-label="item-description"
              class="w-100 p-0"
            >
              <span
                class="h6 v-top "
              >
                Install Angular CLI
              </span>
            </div>
          </div>
          <div
            class="col-auto mt-0"
          >
            <button
              aria-label="btn-remove"
              class="btn btn-danger align-top"
            >
              remove
            </button>
          </div>
        </div>
      </div>
      <hr />
      <div
        aria-label="todo-item"
        class="container-fluid todo-item"
      >
        <div
          class="row"
        >
          <div
            class="col-auto mt-0  v-top"
          >
            <button
              aria-label="btn-todo"
              class="btn btn-block btn-outline-primary item-btn"
            >
              <i
                class="fa-solid fa-check"
              />
            </button>
          </div>
          <div
            class="col-auto mt-0  v-top"
          >
            <button
              aria-label="btn-todo"
              class="btn btn-block btn-primary item-btn"
            >
              <i
                class="fa-solid fa-exclamation"
              />
            </button>
          </div>
          <div
            class="col"
          >
            <div
              aria-label="item-description"
              class="w-100 p-0"
            >
              <span
                class="h6 v-top "
              >
                Create new app
              </span>
            </div>
          </div>
          <div
            class="col-auto mt-0"
          >
            <button
              aria-label="btn-remove"
              class="btn btn-danger align-top"
            >
              remove
            </button>
          </div>
        </div>
      </div>
      <hr />
      <div
        aria-label="todo-item"
        class="container-fluid todo-item"
      >
        <div
          class="row"
        >
          <div
            class="col-auto mt-0  v-top"
          >
            <button
              aria-label="btn-todo"
              class="btn btn-block btn-outline-primary item-btn"
            >
              <i
                class="fa-solid fa-check"
              />
            </button>
          </div>
          <div
            class="col-auto mt-0  v-top"
          >
            <button
              aria-label="btn-todo"
              class="btn btn-block btn-outline-primary item-btn"
            >
              <i
                class="fa-solid fa-exclamation"
              />
            </button>
          </div>
          <div
            class="col"
          >
            <div
              aria-label="item-description"
              class="w-100 p-0"
            >
              <span
                class="h6 v-top "
              >
                Serve app
              </span>
            </div>
          </div>
          <div
            class="col-auto mt-0"
          >
            <button
              aria-label="btn-remove"
              class="btn btn-danger align-top"
            >
              remove
            </button>
          </div>
        </div>
      </div>
      <hr />
      <div
        aria-label="todo-item"
        class="container-fluid todo-item"
      >
        <div
          class="row"
        >
          <div
            class="col-auto mt-0  v-top"
          >
            <button
              aria-label="btn-todo"
              class="btn btn-block btn-outline-primary item-btn"
            >
              <i
                class="fa-solid fa-check"
              />
            </button>
          </div>
          <div
            class="col-auto mt-0  v-top"
          >
            <button
              aria-label="btn-todo"
              class="btn btn-block btn-outline-primary item-btn"
            >
              <i
                class="fa-solid fa-exclamation"
              />
            </button>
          </div>
          <div
            class="col"
          >
            <div
              aria-label="item-description"
              class="w-100 p-0"
            >
              <span
                class="h6 v-top "
              >
                Develop app
              </span>
            </div>
          </div>
          <div
            class="col-auto mt-0"
          >
            <button
              aria-label="btn-remove"
              class="btn btn-danger align-top"
            >
              remove
            </button>
          </div>
        </div>
      </div>
      <hr />
      <div
        aria-label="todo-item"
        class="container-fluid todo-item"
      >
        <div
          class="row"
        >
          <div
            class="col-auto mt-0  v-top"
          >
            <button
              aria-label="btn-todo"
              class="btn btn-block btn-outline-primary item-btn"
            >
              <i
                class="fa-solid fa-check"
              />
            </button>
          </div>
          <div
            class="col-auto mt-0  v-top"
          >
            <button
              aria-label="btn-todo"
              class="btn btn-block btn-outline-primary item-btn"
            >
              <i
                class="fa-solid fa-exclamation"
              />
            </button>
          </div>
          <div
            class="col"
          >
            <div
              aria-label="item-description"
              class="w-100 p-0"
            >
              <span
                class="h6 v-top "
              >
                Deploy app
              </span>
            </div>
          </div>
          <div
            class="col-auto mt-0"
          >
            <button
              aria-label="btn-remove"
              class="btn btn-danger align-top"
            >
              remove
            </button>
          </div>
        </div>
      </div>
    </ul>
  </div>
</div>
`;
